### ISSUE TITLE
Prevent users from staking on machines that cannot calculate proofs in time

### ISSUE THREAD

**Issue by crtahlin** [ethersphere/bee#4849]
#### Summary

Some machines are too slow to calculate proofs required for storage incentive lottery, in time. In that case, although they are properly staking and might have all the data, they are failing to participate fruitfully in the lottery.

There is a way to check the performance of the machines, as described in docs with [rchash](https://docs.ethswarm.org/docs/bee/working-with-bee/bee-api#rchash). But user's might not know they can/should do that. Either they miss it in the docs, or they are using some other method of installation (e.g. DAppNode). 

The proposed new feature would be to required the users to run at least one rchash calculation on a fully synced node as a prerequisite, before being allowed to stake. Which could be overridden by the users, allowing them to stake even if thi prerequisite is not met.  

#### Motivation

A user might run a full staking Bee node, investing time and resources (stake, electricity, hardware, time) and not winning anything, while not even knowing why - that their resources might be insufficient (CPU speed, storage speed). 

#### Implementation

A full Bee node already knows what the result of last rchash calculation was - how long it took to calculate:
![image](https://github.com/user-attachments/assets/3468570d-be7d-4a2c-9ec1-7e4d69c7b350)

The same, or another field, could be used to store the duration of a rchash calculation. Whichever field is used, its value should be retrieved when a staking call is executed, and if the duration is too long, the staking should fail with an informative message, that the machine would probably not be able to play the lottery. If there is no information in the field, the staking should fail, with an informative message, instructing the user to run rchash.

The user can call the staking endpoint with a special argument, overriding the above mentioned check of rchash duration. 

Also, when running rchash call, the returned information should also include a user friendly field, explicitly stating that the calculation was done in required time needed to play in the lottery (perhaps this specific field could be referenced by the logic allowing staking). 

#### Drawbacks

It could affect the onboarding UX negatively, esp. if done improperly - if stopping the user in the onboarding process without proper explanation.

**Comment by n00b21337** [comment]
yes, I missed it as it's just an endpoint that wasn't of particular interest at the time. I only realised that yesterday, when it was brought up that nodes, could miss a lottery. should be in docs with some highest warning on staking part. @NoahMaizels 

think it would be also useful to make this a more rigorous feature, that nodes are tested periodically and output is saved to node and displayed on "status" or something, with average and latest, worst time etc. Why?
because from my example I run it on server with different things on it, and it affects performances, the more other things I install the less resources is for bee, so i might start with enough but could end up with not enough. think some kind of monitoring of this would be good way for node operators to have things in check.

**Comment by ldeffenb** [comment]
I would vote against periodically running this rchash check in any internal fashion.   If you've ever watched the host performance when a node is calculating the hash, it's a very noticeable load on both the CPU and storage subsystems (read: SSD).   If two co-resident nodes in different neighborhoods happen to do an rchash check at the same time, they'll both be longer than they should be, possibly to the point of reporting themselves as "bad" or whatever the intent is for too-long rchash times.

Outside-the-swarm variable host load is one thing, but intentionally adding such a load (random rchash calculation) on every single swarm node is unnecessary, gains nothing, and wastes resources IMHO.

**Comment by n00b21337** [comment]
also make human readable outout of duration currently its this nanoseconds

![image](https://github.com/user-attachments/assets/3bbdc443-c359-4898-bec1-89b7bf8232d1)

**Comment by CanutoIX** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by crtahlin** [comment]
One consideration that affects the design of the gate proposed here: sampling time depends on how full the reserve is, and that is not something the operator controls.

In practice a node holding around 2M chunks within radius (a practical lower bound) computes the reserve commitment hash in roughly half the time of one holding around 4M (the practical upper bound). This is consistent with the implementation — `pkg/storer/sample.go` iterates every chunk in radius via `IterateChunksItems`, so the work is linear in chunk count. Reserve occupancy at any given moment is a property of the network at that time, not a setting.

The consequence for this issue: a gate that records a single `/rchash` duration and then permits staking can issue a false pass — the exact failure it exists to prevent — if the measurement happened while the reserve was light. A node that passes comfortably at 2M may be unable to finish a sample at 4M. Note this is distinct from the "fully synced" condition in the original description: a node can be fully synced and still hold only what the network has placed in its radius so far.

Two adjustments would address it without reintroducing the periodic sampling load @ldeffenb objected to:

- Record the chunk count (or storage radius and reserve size) alongside the duration, so a stored result carries the context it was measured in and can be recognised as optimistic or stale.
- Set the threshold with headroom against the phase window rather than as a bare pass/fail, since a measured figure taken at partial occupancy is a lower bound on future cost.

Neither requires forcing extra sampling. A staking node already computes a sample every round it is selected, so durations and their chunk counts can be recorded as they naturally occur — which also gives the running best/worst/average @0xCardiE asked for, without the artificial load.

Related: the same gap exists on the documentation side in ethersphere/bee-docs#835, where the advice to test with `/rchash` before participating does not mention that the result is only valid for the reserve size at measurement time.

*Description generated with help of AI.*
